src/components/DnsSettings.tsx

```diff
@@ -1,313 +1,239 @@
-    if (!validations) {
-      this.setState({ validateAdditionalRRS: undefined }, this.onUpdate);
-      return;
-    }
-    const validateAdditionalRRS = {
-      failIfMatchesRegexp: validations.failIfMatchesRegexp,
-      failIfNotMatchesRegexp: validations.failIfNotMatchesRegexp,
-    };
-    this.setState({ validateAdditionalRRS }, this.onUpdate);
+    // validation
+    validRCodes: settings!.dns?.validRCodes ?? [DnsResponseCodes.NOERROR],
+    validateAnswerRRS: settings!.dns?.validateAnswerRRS ?? { failIfMatchesRegexp: [], failIfNotMatchesRegexp: [] },
+    validateAuthorityRRS: settings!.dns?.validateAuthorityRRS ?? {
+      failIfMatchesRegexp: [],
+      failIfNotMatchesRegexp: [],
+    },
+    validateAdditionalRRS: settings!.dns?.validateAdditionalRRS ?? {
+      failIfMatchesRegexp: [],
+      failIfNotMatchesRegexp: [],
+    },
   };
-
-  render() {
-    const {
-      ipVersion,
-      recordType,
-      server,
-      port,
-      protocol,
-      showAdvanced,
-      showValidation,
-      validRCodes,
-      validateAnswerRRS,
-      validateAuthorityRRS,
-      validateAdditionalRRS,
-    } = this.state;
-    const { isEditor } = this.props;
-
-    const recordTypes = [
-      {
-        label: DnsRecordType.A,
-        value: DnsRecordType.A,
-      },
-      {
-        label: DnsRecordType.AAAA,
-        value: DnsRecordType.AAAA,
-      },
-      {
-        label: DnsRecordType.CNAME,
-        value: DnsRecordType.CNAME,
-      },
-      {
-        label: DnsRecordType.MX,
-        value: DnsRecordType.MX,
-      },
-      {
-        label: DnsRecordType.NS,
-        value: DnsRecordType.NS,
-      },
-      {
-        label: DnsRecordType.SOA,
-        value: DnsRecordType.SOA,
-      },
-      {
-        label: DnsRecordType.TXT,
-        value: DnsRecordType.TXT,
-      },
-      {
-        label: DnsRecordType.PTR,
-        value: DnsRecordType.PTR,
-      },
-      {
-        label: DnsRecordType.SRV,
-        value: DnsRecordType.SRV,
-      },
-    ];
-    const protocols = [
-      {
-        label: DnsProtocol.UDP,
-        value: DnsProtocol.UDP,
-      },
-      {
-        label: DnsProtocol.TCP,
-        value: DnsProtocol.TCP,
-      },
-    ];
-    const rCodes = [
-      {
-        label: 'NOERROR',
-        value: 'NOERROR',
-      },
-    ];
-    return (
-      <Container>
+  const [dnsSettings, dispatch] = useReducer(dnsSettingsReducer, initialValues);
+  const [showValidation, setShowValidation] = useState(false);
+  const [showAdvanced, setShowAdvanced] = useState(false);
+
+  useEffect(() => {
+    onUpdate({ dns: dnsSettings });
+  }, [dnsSettings]);
+
+  return (
+    <Container>
+      <HorizontalGroup>
+        <Field label="Record Type" description="DNS record type to query for" disabled={!isEditor}>
+          <Select
+            value={dnsSettings.recordType}
+            options={RECORD_TYPES}
+            onChange={(selected: SelectableValue<DnsRecordType>) =>
+              dispatch({
```

Comment:
Using reducers here or any other state management is not necessary since our Form component (as a wrapper for [react-hook-form](https://react-hook-form.com/v5/api) gives that for free. Here's an example of recent code used in [Enterprise](https://github.com/grafana/grafana-enterprise/blob/master/src/public/reports/ReportPage.tsx#L199). Using `register` with `InputControl` eliminates 99% of use cases for custom form management. Takes a bit of practice to get used to, but in the end makes components way simpler.

Probably the biggest challenge here would be to make it work with the `onUpdate` prop since I assume it has to be called every time a field is updated vs once on form submit.

Reply:
That looks like it'll be a major improvement. I'm going to tackle that in it's own PR since it will involve quite a few changes.